app/src/main/java/org/mozilla/fenix/components/metrics/ActivationPing.kt

```diff
@@ -0,0 +1,161 @@
+            null
+        } catch (e: IOException) {
+            Logger.debug("ActivationPing - unable to connect to Google Play Services")
+            null
+        } catch (e: NullPointerException) {
+            Logger.debug("ActivationPing - no Google Advertising ID available")
+            null
+        }
+    }
+
+    /**
+     * Get the salt to use for hashing. This is a convenience
+     * function to help with unit tests.
+     */
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun getHashingSalt(): String = context.applicationInfo.packageName
+
+    /**
+     * Produces an hashed version of the Google Advertising ID.
+     * We want users using more than one of our products to report a different
+     * ID in each of them. This function runs off the main thread and is CPU-bound.
+     *
+     * @return an hashed and salted Google Advertising ID or null if it was not possible
+     *         to get the Google Advertising ID.
+     */
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal suspend fun getHashedIdentifier(): String? = withContext(Dispatchers.Default) {
+        getAdvertisingID()?.let { unhashedID ->
+            // Add some salt to the ID, before hashing. For this specific use-case, it's ok
+            // to use the same salt value for all the hashes. We want hashes to be stable
+            // within a single product, but we don't want hashes to be the same across different
+            // products (e.g. Fennec vs Fenix).
+            val salt = getHashingSalt()
+
+            // Apply hashing.
+            try {
+                val saltedID = unhashedID + salt
+                val digest = MessageDigest.getInstance("SHA-256")
+                val hashedBytes = digest.digest(saltedID.toByteArray(Charsets.UTF_8))
+                Base64.encodeToString(hashedBytes, Base64.NO_WRAP)
+            } catch (e: NoSuchAlgorithmException) {
+                Logger.debug("ActivationPing - no SHA-256 available")
+                null
+            }
+        }
+    }
+
+    /**
+     * Fills the metrics and triggers the 'activation' ping.
+     * This is a separate function to simplify unit-testing.
+     */
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun triggerPing() {
+        // Generate the activation_id.
+        Activation.activationId.generateAndSet()
+
+        CoroutineScope(Dispatchers.Default).launch {
+            val hashedId = getHashedIdentifier()
+            if (hashedId != null) {
+                Logger.info("ActivationPing - generating ping with the hashed id")
+                // We have a valid, hashed Google Advertising ID.
+                Activation.identifier.set(hashedId)
+            }
+
+            Logger.info("ActivationPing - generating ping (has `identifier`: ${hashedId != null})")
+            Pings.activation.send()
+            markAsTriggered()
```

Comment:
Maybe move this to the top of the launch block to avoid the (unlikely) race condition where calling `checkAndSend` twice quickly could send two pings?

Reply:
I'd prefer leaving this as it is: if, for some weird reason, we crash/get killed after that line gets executed, we'd never re-attempt to send the ping. We can easily detect dupes via the hash, on the other hand. So two pings shouldn't be a big deal in this specific case.